cranelift/isle/isle/src/serialize.rs

```diff
@@ -636,82 +661,80 @@ impl<'a> Decomposition<'a> {
-                // Final tie-breaker: prefer constraints on the earliest
-                // binding site.
-                candidate.kind,
-            )
+        // Set aside candidates which currently can't handle any rules. They
+        // can't influence this round, but we'll revisit them later.
+        let progress = partition_in_place(&mut self.candidates, |candidate| candidate.count > 0);
+        let candidates = &mut self.candidates[..progress];
+
+        // Now separate out the equality constraint candidates, which require
+        // special handling, from the normal cases.
+        let normals = partition_in_place(candidates, |candidate| {
+            !matches!(candidate.kind, HasControlFlow::Equal(_, _))
         });
-        let unmatched = self
-            .candidates
-            .partition_point(|candidate| candidate.state != BindingState::Matched);
-        let (mut unmatched, matched) = self.candidates.split_at(unmatched);
-
-        // All valid candidates sort greater than this so the kind of control flow
-        // doesn't matter.
-        let mut best = Candidate::with_invalid_key(HasControlFlow::Match(0.try_into().unwrap()));
-
-        // We sorted the candidates in reverse order of their sort key, so the
-        // first one is usually the best option we have. However, equality-
-        // constraint candidates are over-approximated. Those identify how many
-        // rules have some kind of equality constraint involving a particular
-        // binding site, but we actually need to find the rules with an
-        // equality constraint on a particular _pair_ of binding sites. Those
-        // are the intersection of two candidates' rules, so the upper bound on
-        // the number of matching rules is whichever candidate is smaller.
-        while let Some((candidate, rest)) = unmatched.split_first() {
-            unmatched = rest;
-            if candidate.key() <= best.key() {
+        let (normals, equals) = candidates.split_at_mut(normals);
+
+        // Find the best normal candidate.
+        let mut best = None;
+        for candidate in normals.iter() {
+            if best.as_ref() < Some(candidate) {
+                best = Some(candidate.clone());
+            }
+        }
+
+        // Everything before this took O(n) time.
+
+        // Equality-constraint candidates are over-approximated. They identify
+        // how many rules have some kind of equality constraint involving a
+        // particular binding site, but we actually need to find the rules
+        // with an equality constraint on a particular _pair_ of binding
+        // sites. Those are the intersection of two candidates' rules, so the
+        // upper bound on the number of matching rules is whichever candidate
+        // is smaller. Do an O(n log n) sort before the O(n^2) all-pairs loop
+        // so we can do branch-and-bound style pruning.
+        equals.sort_unstable_by(|x, y| y.cmp(x));
```

Comment:
Is this just a reverse sort...? Could we use `Reverse` if so (`sort_unstable_by_key(|x| Reverse(*x))` or somesuch)?

Reply:
Fair enough; just a comment "note that `x` and `y` are reversed below, to sort in reverse" is probably enough then, IMHO.

Reply:
Sure thing! I've added that comment locally now. I don't feel like making CI run again for a comment change so I'll push it after building the translation verifier we talked about.